app/services/course/assessment/answer/text_response_auto_grading_service.rb

```diff
@@ -77,13 +84,271 @@ def explanations_for(solutions)
   end
 
-  # Mark the correctness of the answer based on solutions
+  # Mark the correctness of the answer based on solutions.
   #
   # @param [Course::Assessment::Question::TextResponse] question The question answered by the
   #   student.
   # @param [Array<Course::Assessment::Question::TextResponseSolution>] solutions The solutions that
   #   matches the student's answer.
+  # @param [Boolean] correct True if the answer is correct.
   def correctness_for(question, solutions)
     solutions.map(&:grade).sum >= question.maximum_grade
   end
+
+  # Grades the given answer for comprehension questions.
+  #
+  # @param [Course::Assessment::Answer::TextResponse] answer The answer specified by the
+  #   student.
+  # @return [Array<(Boolean, Integer, Object)>] The correct status, grade and the messages to be
+  #   assigned to the grading.
+  def evaluate_answer_comprehension(answer)
+    question = answer.question.actable
+    answer_text = answer.normalized_answer_text
+    answer_text_array = answer_text.downcase.gsub(/([^a-z ])/, ' ').split(' ')
+    answer_text_lemma_array = []
+    answer_text_array.each { |a| answer_text_lemma_array.push(WordNet::Synset.morphy_all(a).first || a) }
+
+    answer_text_lemma_status = { 'compre_lifted_word' => Array.new(answer_text_lemma_array.length, nil),
+                                 'compre_keyword' => Array.new(answer_text_lemma_array.length, nil) }
+
+    hash_lifted_word_points = hash_compre_lifted_word(question)
+
+    hash_keyword_solutions = hash_compre_keyword(question)
+
+    find_compre_lifted_word_in_answer(answer_text_lemma_array,
+                                      answer_text_lemma_status,
+                                      hash_lifted_word_points)
+
+    find_compre_keyword_in_answer(answer_text_lemma_array,
+                                  answer_text_lemma_status,
+                                  hash_keyword_solutions)
+
+    answer_grade = grade_for_comprehension(question, answer_text_lemma_status)
+
+    [
+      correctness_for_comprehension(question, answer_grade),
+      answer_grade,
+      explanations_for_comprehension(question, answer_grade, answer_text_array, answer_text_lemma_status)
+    ]
+  end
+
+  # All lifted words in a question as keys and
+  # an array of Points where words are found as values, for comprehension questions.
+  #
+  # @param [Course::Assessment::Question::TextResponse] question The question answered by the
+  #   student.
+  # @return [Hash{String=>Array<Course::Assessment::Question::TextResponseComprehensionPoint>}]
+  #   The mapping from lifted words to Points.
+  def hash_compre_lifted_word(question)
+    hash = {}
+    question.groups.each do |group|
+      group.points.each do |point|
+        # for all TextResponseComprehensionSolution where solution_type == compre_lifted_word
+        point.solutions.select(&:compre_lifted_word?).each do |s|
+          s.solution_lemma.each do |solution_key|
+            if hash.key solution_key
```

Comment:
Did you mean `key?`?

Also, our current style convention is not to omit the parentheses unless we are using a DSL. Same for `push` and `include?` below.

Reply:
Oops, miss out the `?` in `key?`. I have also included the parentheses.